content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+the includes so we do not have these seemingly isolated files.
+
+# Dump
+
+What about printf? The best debugging tool of them all.
+
+TableGen’s equivalent is
+[dump](https://llvm.org/docs/TableGen/ProgRef.html#dump-print-messages-to-stderr),
+and its companion `repr`.
+```
+class Register<int _size, string _alias=""> {
+   int size = _size;
+   string alias = _alias;
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
+}
+
+def X0: Register<8> {}
+def X29: Register<8, "frame pointer"> {}
+```
+
+`dump` prints to `stderr`:
+```
+<source>:4:5: note: size is 8 and alias is ""
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
+   ^
+<source>:4:5: note: size is 8 and alias is "frame pointer"
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
+   ^
+```
+
+Note: This was added
+[recently](https://github.com/llvm/llvm-project/commit/411c4edeef076bd2e01b104fe095ba381600a3d3)
+by Francesco Petrogalli and Adam Nemet. So you will need a recent build, or a
+released version 18.0 or greater (which is unreleased at time of writing).
+
+Of course you can try this
+[on Compiler Explorer](https://godbolt.org/z/555n19qMK) right now!
+
+# Find In Files
+
+This is last because in an ideal world it would be the last option, but it’s
+often not the least of them. Grep, ack, Find In Files, whatever you call it,
+searching text is unreasonably effective if you have a little knowledge of the
+language syntax.
+
+Should I even mention such an obvious idea? Well obvious is in the eye of the
+beholder, and for TableGen we have a special circumstance that makes it more
+effective than usual.
+
+In the LLVM project repo we have the vast majority of TableGen code in use today.
+Want to know how to use a feature? It’s all there, albeit in a 500,000+ line
+haystack, but you’d be surprised what a simple query can find.
+
+Think about how you would write the thing you’re trying to find. If it’s a
+class would it have template arguments or not and so would there be a `<` after
+the name? If it’s an error message, what parts would be constant and what parts
+substituted?
+
+`class Foo has no attribute Bar` is much more likely to be created by
+substituting the name of the class and attribute whereas `Expected end of line`
+is likely a static string.
+
+There are also tests for the compiler, most of which are in
+[this folder](https://github.com/llvm/llvm-project/tree/main/llvm/test/TableGen).
+Where you will find minimal examples for the language features. You could
+narrow your search to that location.
+
+# Conclusion
+
+Learning TableGen doesn’t have to be scary. Don’t think that because it’s an
+isolated DSL that it doesn’t have what you’ve come to expect from your
+favourite languages.
+
+Keep in mind that TableGen is also a tool, not a goal unto itself. If you can
+achieve your goals with a limited but accurate understanding of TableGen and its
+backends, that’s fine. Dive as deep as you need or want to.
```

Comment:
Dive as deep as you want or need.

Reply:
I don't know that "dive" counts as an idiom, but I'll err on the side of caution. So it's now "Learn as much..." but I've added your suggestion to that.

Which is maybe a bit awkward to read, but I can't tell.